### PR TITLE
Add an agent plugin to support elasticsearch7 

### PR DESCRIPTION
**Pull request by zifeihan**
### Add an agent plugin to support elasticsearch7 
- [x] Add a test case for the new plugin, refer to [the doc](https://github.com/apache/skywalking/blob/master/docs/en/guides/Plugin-test.md)
- [x] Add a component id in [the component-libraries.yml](https://github.com/apache/skywalking/blob/master/oap-server/server-bootstrap/src/main/resources/component-libraries.yml)
- [ ] Add a logo in [the UI repo](https://github.com/apache/skywalking-rocketbot-ui/tree/master/src/views/components/topology/assets)
    
Hello, I found that the current es6 plugin is not applicable to the transport-related api on es7, not only the transport api, so the es7 plugin is used to process the related logic of es7. The following exceptions may occur when using the original es6 plugin, and it will cause span not to exit normally, and threadLocal will not be cleaned up, log message such as
```
{
  "@timestamp": "2021-04-08 03:53:23:984",
  "agent_name": "teachers@10278",
  "level": "ERROR",
  "throwable": "\njava.lang.NoSuchMethodError: org.elasticsearch.search.SearchHits.getTotalHits()J\n\tat org.apache.skywalking.apm.plugin.elasticsearch.v6.interceptor.AdapterActionFutureActionGetMethodsInterceptor.parseSearchResponse(AdapterActionFutureActionGetMethodsInterceptor.java:118)\n\tat org.apache.skywalking.apm.plugin.elasticsearch.v6.interceptor.AdapterActionFutureActionGetMethodsInterceptor.parseResponseInfo(AdapterActionFutureActionGetMethodsInterceptor.java:86)\n\tat org.apache.skywalking.apm.plugin.elasticsearch.v6.interceptor.AdapterActionFutureActionGetMethodsInterceptor.afterMethod(AdapterActionFutureActionGetMethodsInterceptor.java:66)\n\tat org.apache.skywalking.apm.agent.core.plugin.interceptor.enhance.InstMethodsInter.intercept(InstMethodsInter.java:102)\n\tat org.elasticsearch.action.support.AdapterActionFuture.actionGet(AdapterActionFuture.java)\n\tat org.elasticsearch.action.ActionRequestBuilder.get(ActionRequestBuilder.java:52)\n\tat com.zhangmen.teacher.service.internal.impl.TeacherArrangeServiceImpl.getBatchTeaSimpleInfoByEs(TeacherArrangeServiceImpl.java:224)\n\tat com.zhangmen.teacher.service.internal.impl.TeacherArrangeServiceImpl$$FastClassBySpringCGLIB$$e5aec092.invoke(<generated>)\n\tat org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)\n\tat org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:738)\n\tat org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)\n\tat org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:85)\n\tat java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)\n\tat org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)\n\tat java.lang.Thread.run(Thread.java:748)\n",
  "logger": "InstMethodsInter",
  "thread": "http-nio-8080-exec-7",
  "message": "class[class org.elasticsearch.action.support.PlainActionFuture] after method[actionGet] intercept failure"
}
```

### CODE REVIEW

**Comment by wu-sheng** [comment]
> Hello, I found that the current es6 plugin is not applicable to the transport-related api on es7, not only the transport api, so the es7 plugin is used to process the related logic of es7. 

You missed the update of supported-list.md

**Comment by wu-sheng** [comment]
@zifeihan Please check CI

**Comment by zifeihan** [comment]
> @zifeihan Please check CI

done.

**Comment by zifeihan** [comment]
> > Hello, I found that the current es6 plugin is not applicable to the transport-related api on es7, not only the transport api, so the es7 plugin is used to process the related logic of es7.
> 
> You missed the update of supported-list.md

fixed.

**Comment by dmsolr** [comment]
I just took a look and it is almost ok. Let me have a look tomorrow. :)

**Comment by wu-sheng** [comment]
> --- /home/runner/work/skywalking/skywalking/tools/plugin/md-javaagent-plugin-list.txt	2021-04-16 00:06:27.776825759 +0000
+++ /home/runner/work/skywalking/skywalking/tools/plugin/genernate-javaagent-plugin-list.txt	2021-04-16 00:06:27.772825656 +0000
@@ -18,0 +19 @@
+elasticsearch-7.x

Please check the CI task, you add a new plugin.

File: apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/define/AdapterActionFutureInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+    @Override
+    protected List<WitnessMethod> witnessMethods() {
+        return Arrays.asList(new WitnessMethod[] {
+            new WitnessMethod(
+                Constants.SEARCH_HITS_WITNESS_CLASSES,
+                named("getTotalHits").and(takesArguments(0)).and(returns(named("org.apache.lucene.search.TotalHits")))
+            )
+        });
+    }
```
Comment (dmsolr) [inline apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/define/AdapterActionFutureInstrumentation.java:90]: I am thinking about whether we really need to add a separate plugin for ES7. The problem cloud is resolved by using an IF statement here. Right?

Comment (zifeihan) [reply apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/define/AdapterActionFutureInstrumentation.java:90]: If add es7-related judgment logic in the es6 plugin, it feels a bit strange. and adding es7 plugins will allow us to do more es7-related enhancements in the future.

File: apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/interceptor/AdapterActionFutureActionGetMethodsInterceptor.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        Tags.DB_TYPE.set(span, Constants.DB_TYPE);
+    }
+
+    @Override
+    public Object afterMethod(EnhancedInstance objInst, Method method, Object[] allArguments,
+                              Class<?>[] argumentsTypes, Object ret) throws Throwable {
+
+        if (!isTrace(objInst)) {
+            return ret;
+        }
```
Comment (dmsolr) [inline apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/interceptor/AdapterActionFutureActionGetMethodsInterceptor.java:63]: Maybe we invert the condition. Change it to `isNotTrace(...)`

File: apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/define/AdapterActionFutureInstrumentation.java (hunk 1 above)
Comment (dmsolr) [reply apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/define/AdapterActionFutureInstrumentation.java:90]: Make sense.

Comment (dmsolr) [reply apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/define/AdapterActionFutureInstrumentation.java:90]: I have a code style suggestion:
```java
@Override
protected List<WitnessMethod> witnessMethods() {
    return Collections.singletonList(new WitnessMethod(
            Constants.SEARCH_HITS_WITNESS_CLASSES,
            named("getTotalHits").and(takesArguments(0)).and(returns(named("org.apache.lucene.search.TotalHits")))
    ));
}
```

File: apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/interceptor/AdapterActionFutureActionGetMethodsInterceptor.java (hunk 2 above)
Comment (zifeihan) [reply apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/interceptor/AdapterActionFutureActionGetMethodsInterceptor.java]: done.

File: apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/define/AdapterActionFutureInstrumentation.java (hunk 1 above)
Comment (zifeihan) [reply apm-sniffer/apm-sdk-plugin/elasticsearch-7.x-plugin/src/main/java/org/apache/skywalking/apm/plugin/elasticsearch/v7/define/AdapterActionFutureInstrumentation.java]: thank you, this has been fixed, please help recheck.

**Review by dmsolr — APPROVED** [review]
LGTM. Let us wait for CI to pass.

**Pull request merged**